mono/utils/mono-threads-windows.c

```diff
@@ -391,6 +442,61 @@ mono_native_thread_set_name (MonoNativeThreadId tid, const char *name)
 	}
 	__except(EXCEPTION_EXECUTE_HANDLER) {
 	}
+
+#endif
+
+#if G_HAVE_API_SUPPORT(HAVE_CLASSIC_WINAPI_SUPPORT)
+	/* UWP does not allow LoadLibrary. */
+
+	/* Bing: SetThreadDescription.
+	 * https://msdn.microsoft.com/en-us/library/windows/desktop/mt774976(v=vs.85).aspx
+	 * This mechanism has the following advantages over the older mechanism:
+	 * - Supported by the kernel, not debuggers.
+	 * - It works when debugger is not attached.
+	 *   - Either before debugger is attached.
+	 *   - Between detach and reattach.
+	 *   - No debugger at all.
+	 *   - This is arguably a memory footprint disadvantage.
+	 * - There is an API to get the name, instead of being buried in the debugger.
+	 * - ETW knows about it, so thread names appear in ETW traces.
+	 * - Unicode (only) support.
+	 * - Implemented by Mono team.
+	 * - No exception used to implement it -- compiler-portable.
+	 * - It works on handles instead of IDs, which is more idiomatic.
+	 *
+	 * It has the following disadvantages:
+	 * - It is only present in newer versions of Windows -- requiring LoadLibrary/GetProcAddress.
+	 *   LoadLibrary/GetProcAddress are not allowed in UWP.
+	 *   You are supposed to delayload and check if delayload succeeded,
+	 *   however such requirement from a static lib to its client is onerous.
+	 * - The Xbox Win32 API is different, though the underlying kernel API is ABI-identical.
+	 * - The Xbox360 API is another different.
+	 *
+	 * It is called "description" instead of "name" because SetThreadName is already taken.
+	 * It was requested by Bruce Dawson of Google, formerly of Microsoft/Xbox.
+	 * https://randomascii.wordpress.com/2015/10/26/thread-naming-in-windows-time-for-something-better */
+	typedef HRESULT (WINAPI *PFN_SET_THREAD_DESCRIPTION) (HANDLE, PCWSTR);
+	static PFN_SET_THREAD_DESCRIPTION setThreadDescription;
+	static const Kernel32Proc proc = { .name = "SetThreadDescription" };
+
+	if (!proc.sticky_fail) {
+		*(PROC*)&setThreadDescription = get_kernel32_proc_address (&proc);
+
+		if (setThreadDescription) {
+			HANDLE handle;
+			gunichar2 *name_utf16;
+
+			handle = OpenThread (THREAD_ALL_ACCESS, FALSE, tid);
+			g_assert (handle);
+
+			name_utf16 = g_utf8_to_utf16 (name, strlen (name), NULL, NULL, NULL);
```

Comment:
This part I don't want in the eventual result. Managed already has a nul terminated string. It can flow through w/o copy or conversion. I was deliberating waiting on this until other threads.c changes are done.

Reply:
Also, setting this on self is less code than setting it on arbitrary -- you can use GetCurrentThread.
The bulk of this change is correct long term, so maybe that is a reason to take it sooner rather than later. But it will/should change again.